src/ledger/grain.js

```diff
@@ -185,4 +194,87 @@ export function toFloatRatio(numerator: Grain, denominator: Grain): number {
   return Number(numerator) / Number(denominator);
 }
 
+/**
+ * Splits a budget of Grain into fractional shares.
+ *
+ * It guarantees that the sum of all the splits will be exactly equal
+ * to the budget. It does not guarantee fairness, i.e. some users may
+ * receive more or less Grain than others with equal shares.
+ */
+export function splitBudget(
+  budget: Grain,
+  shares: $ReadOnlyArray<number>
+): $ReadOnlyArray<Grain> {
+  if (lt(budget, ZERO)) {
+    throw new Error("negative budget");
+  }
+  let remainingBudget = budget;
+  let remainingShares = shares.reduce((a, b) => a + b, 0);
+  if (remainingShares === 0) {
+    throw new Error("shares sum to 0");
+  }
+  if (!isFinite(remainingShares)) {
+    throw new Error(`shares sum not finite: ${remainingShares}`);
+  }
+  const pieces = shares.map((s) => {
+    if (s < 0) {
+      throw new Error("negative share: " + s);
+    }
+    if (s === 0 || remainingShares === 0) {
+      // You would think that s === 0 implies remainingShares === 0, but
+      // testing in extreme circumstances reveals that both checks are needed.
+      return "0";
+    }
+    let fraction = s / remainingShares;
+    if (fraction > 1) {
+      fraction = 1;
+    }
+    let piece = multiplyFloat(remainingBudget, fraction);
+
+    /**
+     * Uncomment below if you want to measure the discrepancy caused by
+     * forcing fracion=1 whenever fraction > 1.
+     *
+     * In testing, when run with wildly varying shares, it never produced more than
+     * ~hundreds of attoGrain discrepancy.
+     */
+    /*
+    const altPiece = multiplyFloat(remainingBudget, s / remainingShares);
+    if (altPiece !== piece) {
+      const delta = sub(altPiece, piece);
+      console.error(
+        `${delta} discrepancy due to capping ${s} / ${remainingShares} from ${
+          s / remainingShares
+        } to 1`
+      );
+    }
+    */
+
+    if (piece === "0" && remainingBudget !== "0" && s > 0) {
+      piece = "1";
+    }
+    remainingBudget = sub(remainingBudget, piece);
+    remainingShares -= s;
+    return piece;
+  });
+  // istanbul ignore if
+  if (lt(remainingBudget, "0")) {
```

Comment:
Would be good to have a comment explaining why we are not testing for coverage here

Reply:
It's an invariant checker--if we found a test case that triggered this case, then it would mean the function is broken, so we can't expect to hit it in testing.

Reply:
Added a comment.